cron/uptycs_alertpull.py

```diff
@@ -0,0 +1,288 @@
+    #     del (normalized["user"])
+
+    # return normalized
+    return details
+
+
+def process_alerts(mozmsg, uptycs_alerts):
+    for alert in uptycs_alerts:
+        details = {}
+        # Timestamp format: http://mozdef.readthedocs.io/en/latest/usage.html#mandatory-fields
+        # Duo logs come as a UTC timestamp
+        dt = datetime.utcfromtimestamp(e["timestamp"])
+        mozmsg.timestamp = dt.replace(tzinfo=utc).isoformat()
+
+        # TODO: Fix this so hostName is properly set
+        mozmsg.log["hostname"] = alert["host"]
+        for item in alert:
+            details[item] = alert[item]
+
+        localdetails = normalize(details)
+        mozmsg.details = localdetails
+        mozmsg.summary = (
+            alert["eventtype"] + " " + alert["result"] + " for " + alert["username"]
+        )
+        mozmsg.send()
+
+
+def main():
+    mozmsg = mozdef.MozDefEvent(options.MOZDEF_URL)
+    mozmsg.tags = ["uptycs"]
+    mozmsg.set_category("uptycs")
+    mozmsg.source = "UptycsAPI"
+    if options.DEBUG:
+        mozmsg.debug = options.DEBUG
+        mozmsg.set_send_to_syslog(True, only_syslog=True)
+
+    client = UptycsClient(options.UPTYCS_API_JSON_FILE)
+
+    # Adjust the filter as needed to set proper search window
+    filter = UptycsFilter()
+    utc_now = datetime.datetime.utcnow()
+
+    # If an existing state file exists, set our start window to last run
+    if os.path.exists(options.statepath):
+        state = pickle.load(open(options.statepath, 'rb'))
+        filter.start_time = state["last_run"]
+        last_alert_ids = state["last_alert_ids"]
+    else:
+        last_alert_ids = []
+        filter.start_time = utc_now - datetime.timedelta(days=30)
+
+    filter.end_time = utc_now
+
+    # Query alerts that match the filter
+    alerts = client.alerts(filter)
+    alert_ids = []
+    if len(alerts) > 0:
+        for alert in alerts:
+            if alert['id'] in last_alert_ids:
+                continue
+            else:
+                alert_ids.append(alert['id'])
+
+    # Process all these alerts in MozDef
+    process_alerts(mozmsg, alerts)
+
+    state = {
+        "last_run": filter.end_time,
+        "last_alert_ids": alert_ids,
+    }
+    pickle.dump(state, open("poc.state", "wb"))
+
+
+def initConfig():
+    options.UPTYCS_API_JSON_FILE = getConfig("UPTYCS_API_JSON_FILE", "", options.configfile)
+    options.MOZDEF_URL = getConfig("MOZDEF_URL", "", options.configfile)
+    options.DEBUG = getConfig("DEBUG", True, options.configfile)
+    options.statepath = getConfig("statepath", "", options.configfile)
+
+
+if __name__ == "__main__":
+    parser = OptionParser()
+    defaultconfigfile = sys.argv[0].replace(".py", ".conf")
+    parser.add_option(
+        "-c",
+        dest="configfile",
+        default=defaultconfigfile,
+        help="configuration file to use",
+    )
+    (options, args) = parser.parse_args()
+    initConfig()
+    main()
+
+
+# Example Alert JSON
```

Comment:
@pwnbus here's a pretty print alert JSON example from Uptycs, please let me know if there are specific key fields you can immediately grok that we should normalize for and I can make the adjustments in code.

Reply:
I just ran it against our mapping from production and it appears that we don't have any conflicts, so we should be good to go!